source/material_model/composition_reaction.cc

```diff
@@ -63,16 +69,16 @@ namespace aspect
                 break;
             }
 
-          const double c1 = composition.size()>0?
-                            std::max(0.0, composition[0])
-                            :
-                            0.0;
-          const double c2 = composition.size()>1?
-                            std::max(0.0, composition[1])
-                            :
-                            0.0;
-          out.densities[i] = reference_rho * (1 - thermal_alpha * (temperature - reference_T))
-                             + compositional_delta_rho_1 * c1 + compositional_delta_rho_2 * c2;
+          equation_of_state.evaluate(in, i, eos_outputs);
+
+          std::vector<double> volume_fractions (n_compositions_for_eos, 1.0);
+          for (unsigned int c=0; c<n_compositions_for_eos-1; ++c)
+            {
+              volume_fractions[c+1] = std::max(0.0, in.composition[i][c]);
+              volume_fractions[0] -= volume_fractions[c+1];
```

Comment:
do we need to ensure this value stays positive? You could add a max(..) after the loop.

Reply:
We didn't do that in the old logic, and I didn't want to change functionality. If you think we should change it, I can make a separate pull request.